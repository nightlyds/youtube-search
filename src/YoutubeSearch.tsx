/* eslint-disable react/jsx-curly-newline */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useEffect, useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
    faTimesCircle,
    faThumbsUp,
    faThumbsDown,
    faComments,
    faEye,
} from "@fortawesome/free-solid-svg-icons";
import Loading from "./components/Loading";
import Error from "./components/Error";
import Footer from "./components/Footer";

export type VideosTypes = {
    [key: string]: any;
}; // Types for the Video Object

export type YoutubeSearchProps = {
    mostPopular: () => void;
    videos: VideosTypes;
    loading: boolean;
    error: boolean;
    changeKeyword: (keyword: string | undefined) => void;
}; // Types for props

const YoutubeSearch = ({
    mostPopular,
    videos,
    loading,
    error,
    changeKeyword,
}: YoutubeSearchProps) => {
    const [keyword, setKeyword] = useState<string>("");

    useEffect(() => {
        mostPopular();
    }, []); // Load videos before the component did mount

    function clearField(): void {
        setKeyword(""); // Clear input field
        mostPopular(); // Load the Most Popular videos
    }

    return (
        <div className="youtube-search-container">
            <div className="youtube-search-bar">
                <div className="youtube-search-field-box">
                    <input
                        type="text"
                        className="youtube-search-field"
                        value={keyword}
                        onChange={event => {
                            setKeyword(event.target.value);
                        }}
                    />
                    <FontAwesomeIcon
                        icon={faTimesCircle}
                        className="youtube-search-field-icon"
                        onClick={() => clearField()}
                    />
                </div>
                <div className="youtube-search-button-box">
                    <span
                        className="youtube-search-button"
                        onClick={() => changeKeyword(keyword)}
                    >
                        Search
                    </span>
                </div>
            </div>
            <div className="youtube-search-content">
                {loading &&
                    !error &&
                    videos &&
                    videos.data.items.map(
                        (item: VideosTypes, index: number) => (
                            <div
                                className="youtube-search-content-video"
                                key={
                                    item.id.videoId
                                        ? `${item.id.videoId}${index}_`
                                        : `${item.id}_${index}`
                                } // Unique key with the item ID and index of the item
                            >
                                <a
                                    href={`/videos/${`${
                                        item.id.videoId
                                            ? item.id.videoId
                                            : item.id
                                    }`}`}
                                >
                                    <div className="youtube-search-content-video-image-box">
                                        <img
                                            src={
                                                item.snippet.thumbnails.high.url
                                            }
                                            className="youtube-search-content-video-image"
                                            alt="Video cover"
                                        />
                                    </div>
                                    <div className="youtube-search-content-video-description-box">
                                        <h3 className="youtube-search-content-video-description-title">
                                            {item.snippet.title}
                                        </h3>
                                        <span className="youtube-search-content-video-description-channel-title">
                                            {item.snippet.channelTitle}
                                        </span>
                                        {item.statistics && (
                                            <div className="youtube-search-content-video-description-statistics">
                                                <span className="youtube-search-content-video-description-viewers">
                                                    <FontAwesomeIcon
                                                        icon={faEye}
                                                        className="youtube-search-content-video-description-icon"
                                                    />
                                                    {item.statistics.viewCount}
                                                </span>
                                                <span className="youtube-search-content-video-description-likes">
                                                    <FontAwesomeIcon
                                                        icon={faThumbsUp}
                                                        className="youtube-search-content-video-description-icon"
                                                    />
                                                    {item.statistics.likeCount}
                                                </span>
                                                <span className="youtube-search-content-video-description-dislikes">
                                                    <FontAwesomeIcon
                                                        icon={faThumbsDown}
                                                        className="youtube-search-content-video-description-icon"
                                                    />
                                                    {
                                                        item.statistics
                                                            .dislikeCount
                                                    }
                                                </span>
                                                <span className="youtube-search-content-video-description-comments">
                                                    <FontAwesomeIcon
                                                        icon={faComments}
                                                        className="youtube-search-content-video-description-icon"
                                                    />
                                                    {
                                                        item.statistics
                                                            .commentCount
                                                    }
                                                </span>
                                            </div>
                                        )}
                                    </div>
                                </a>
                            </div>
                        )
                    )}
                {!loading && <Loading />}
                {loading && error && <Error />}
            </div>
            <Footer />
        </div>
    );
};

export default YoutubeSearch;
