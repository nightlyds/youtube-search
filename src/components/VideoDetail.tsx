/* eslint-disable react/jsx-one-expression-per-line */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState, useEffect } from "react";
import { RouteComponentProps } from "react-router";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
    faArrowLeft,
    faHeart,
    faThumbsUp,
    faThumbsDown,
    faComments,
    faEye,
} from "@fortawesome/free-solid-svg-icons";
import { VideosTypes } from "../YoutubeSearch";
import Loading from "./Loading";
import Error from "./Error";

type RouteIDType = {
    id: string;
};

export interface VideoDetailProps extends RouteComponentProps<RouteIDType> {
    loadVideoDetail: (id: string) => void;
    addToFavourites: (favourites: Array<string>) => void;
    favourites: Array<string>;
    videoDetail: VideosTypes;
    loading: boolean;
    error: boolean;
}

const VideoDetail = (props: VideoDetailProps) => {
    const {
        match,
        loadVideoDetail,
        videoDetail,
        addToFavourites,
        favourites,
        loading,
        error,
    } = props;
    const [iconFavorite, setIconFavorite] = useState<boolean>(
        favourites.includes(`${match.params.id}`)
    ); // Default value of the video favourite

    useEffect(() => {
        loadVideoDetail(`${match.params.id}`); // Load data for the video
    }, []);

    function addToFavouriteVideo(id: string) {
        setIconFavorite(!iconFavorite);
        // If value is in the array we delete it, else we add it
        if (favourites.includes(id)) {
            const newArray = favourites;
            newArray.splice(newArray.indexOf(id), 1);
            addToFavourites(newArray);
            localStorage.removeItem(`${id}`);
            localStorage.setItem("favourites", JSON.stringify(newArray));
        } else {
            addToFavourites([...favourites, id]);
            localStorage.setItem(
                "favourites",
                JSON.stringify([...favourites, id])
            );
        }
    }
    return (
        <div className="video-detail">
            <div className="video-detail-back-box">
                <a href="/">
                    {" "}
                    <FontAwesomeIcon
                        icon={faArrowLeft}
                        className="video-detail-back-icon"
                    />{" "}
                    Back to Home
                </a>
            </div>
            {loading &&
                !error &&
                videoDetail &&
                videoDetail.data.items.map((item: any) => (
                    <div
                        className="video-detail-content"
                        key={
                            `${item.id}}` // Unique key with the item ID and index of the item
                        }
                    >
                        <div className="video-detail-content-videobox">
                            <iframe
                                src={`https://www.youtube.com/embed/${item.id}`}
                                className="video-detail-content-video"
                                title="Video"
                                frameBorder="0"
                                allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
                                allowFullScreen
                            />
                        </div>
                        <div className="video-detail-content-video-description-box">
                            <h3 className="video-detail-content-video-description-title">
                                {item.snippet.title}
                            </h3>
                            <span className="video-detail-content-video-description-channel-title">
                                {item.snippet.channelTitle}
                            </span>
                            {item.statistics && (
                                <div className="youtube-search-content-video-description-statistics">
                                    <span className="youtube-search-content-video-description-viewers">
                                        <FontAwesomeIcon
                                            icon={faEye}
                                            className="youtube-search-content-video-description-icon"
                                        />
                                        {item.statistics.viewCount}
                                    </span>
                                    <span className="youtube-search-content-video-description-likes">
                                        <FontAwesomeIcon
                                            icon={faThumbsUp}
                                            className="youtube-search-content-video-description-icon"
                                        />
                                        {item.statistics.likeCount}
                                    </span>
                                    <span className="youtube-search-content-video-description-dislikes">
                                        <FontAwesomeIcon
                                            icon={faThumbsDown}
                                            className="youtube-search-content-video-description-icon"
                                        />
                                        {item.statistics.dislikeCount}
                                    </span>
                                    <span className="youtube-search-content-video-description-comments">
                                        <FontAwesomeIcon
                                            icon={faComments}
                                            className="youtube-search-content-video-description-icon"
                                        />
                                        {item.statistics.commentCount}
                                    </span>
                                </div>
                            )}
                            <div className="video-detail-add-to-favourite-box">
                                <span
                                    className="video-detail-add-to-favourite"
                                    onClick={() => addToFavouriteVideo(item.id)}
                                >
                                    Add to Favourite{" "}
                                    <FontAwesomeIcon
                                        icon={faHeart}
                                        className={`video-detail-add-to-favourite-icon ${
                                            iconFavorite
                                                ? "video-detail-add-to-favourite-icon-favorite"
                                                : ""
                                        }`}
                                    />
                                </span>
                            </div>
                        </div>
                    </div>
                ))}
            {!loading && <Loading />}
            {loading && error && <Error />}
        </div>
    );
};

export default VideoDetail;
