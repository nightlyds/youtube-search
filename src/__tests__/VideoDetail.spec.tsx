import React from "react";
import VideoDetail, { VideoDetailProps } from "../components/VideoDetail";
import { VideosTypes } from "../YoutubeSearch";
import { shallow } from "enzyme";

const setUp = (props: VideoDetailProps) => shallow(<VideoDetail {...props} />);

describe("VideoDetail component tests", () => {
    let component: any;
    const loadVideoDetail = (id: string) => {};
    const addToFavourites = (favourites: Array<string>) => {};
    const favourites: Array<string> = [];
    const videoDetail: Array<VideosTypes> = [];
    const loading: boolean = false;
    const error: boolean = false;
    let history: any;
    let match: any = {
        params: {
            id: "ID",
        },
    };
    let location: any;

    beforeEach(() => {
        component = setUp({
            history,
            match,
            location,
            loadVideoDetail,
            addToFavourites,
            favourites,
            videoDetail,
            loading,
            error,
        });
    });

    it("Snapshot test", () => {
        expect(component).toMatchSnapshot();
    });
});
