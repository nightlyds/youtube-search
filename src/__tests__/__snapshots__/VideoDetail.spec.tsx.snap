// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail component tests Snapshot test 1`] = `
<div
  className="video-detail"
>
  <div
    className="video-detail-back-box"
  >
    <a
      href="/"
    >
       
      <FontAwesomeIcon
        border={false}
        className="video-detail-back-icon"
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f060",
              "M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z",
            ],
            "iconName": "arrow-left",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       
      Back to Home
    </a>
  </div>
  <Loading />
</div>
`;
